### ISSUE TITLE
Port already in use fails silently causing red screen of death

### ISSUE THREAD

**Issue by JustinCarmony** [facebook/react-native#257]
I'm on the latest Xcode (6.2), node (v0.10.36), npm(2.7.3), watchman (3.0.0), and flow (0.1.6).

I run the ``react-native init testProject``. I open the testProject.xcodeproj in Xcode and press the run button to run it on iPhone 6 iOS 8.2. I see the first intro screen, then if I click anywhere I immediately get a Red Screen saying:

```
SyntaxError: Unexpected token '<'

{
    "line": 1,
    "sourceURL":"http://localhost:8081/index.ios.bundle"
}
```

![image](https://cloud.githubusercontent.com/assets/602176/6857442/63fd4f0a-d3cc-11e4-871f-875b0c784611.png)

I've tried init'ing multiple projects, all the same effect. It looks like it isn't compiling the JSX correctly? I'm more than willing to help debug, just need to be pointed in the right direction.

**Comment by amasad** [comment]
Can you open http://localhost:8081/index.ios.bundle in your browser and past the result somewhere?

**Comment by JustinCarmony** [comment]
Looks like it is the nginx welcome screen?

```html
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

**Comment by spicyj** [comment]
react-native doesn't do anything with nginx – perhaps you have something else running on that port? Not sure why that wouldn't give an error earlier though…

**Comment by JustinCarmony** [comment]
Let me hunt down what is using that port.

**Comment by JustinCarmony** [comment]
Ok, so I have a Vagrant project running that had opened 8081. When I turned it off, I could validate that I was appropriately getting the bundle file, but I would still get the error. I tried doing a "clean" before running again, same issue. 

I had to delete the app from my simulator, run, and then it worked. :)

It appears that packager was failing silently to run. I didn't know I should see a command terminal pop open, so I had no idea.

So thoughts on how to avoiding in the future:

1) Have packager, if it can't open the port, do something other than silently fail. Would have clued me in much sooner.
2) Allow the user to configure what port to use.

Thanks for the quick response! I'm super excited to build something! :)

I'll leave this open for now and let you close it incase you wanted to make any other issues about the silently failing packager or configuring the ports. :)

**Comment by amasad** [comment]
Yup, it should just fail. Glad it's working for you now.

**Comment by JustinCarmony** [comment]
@amasad want me to re-name the title of the Issue to "Port Already In Use Fails Silently" and we can figure out how to fix that?

**Comment by amasad** [comment]
that'd be great!

**Comment by JustinCarmony** [comment]
@amasad Ok, so I found a snag, not sure how to best solve this.

[``React/React.xcodeproj/project.pbxproj``](https://github.com/facebook/react-native/blob/9c2dbbf99744945f32ca7c3fdbfaba044fdd4d1d/React/React.xcodeproj/project.pbxproj#L448) on Line 448, it checks to see if packager is running or not by checking port 8081 (since people can run the packager on their own). So if you have something else running on 8081, it won't run the packager.

So the question is, how do we check to make sure whatever is running on 8081 is valid?

Some ideas:

1) On react-native-cli when calling init, it could check to see if 8081 is already in use and emit a warning. It wouldn't help if sometime down the road another app listens to 8081, but it could help a brand new user.
2) Put some notice in the Xcode console log that it is not launching packager because 8081 is in use. This would make users aware something is wrong if they are not running packager on their own.
3) Have the react-native somehow validate that it is listening to a proper packager instance (i.e. searching for a string from http://localhost:8081/index.ios.bundle)

I'm not sure which, if any, are the best approach. Any thoughts?

**Comment by amasad** [comment]
I like something like the 3rd option. We can have a status page exposed from the packager that gets `curl`ed in the same line you pointed to.

cc @frantic

**Comment by JustinCarmony** [comment]
@amasad I made a simple pull request with it. Feel free to give any feedback on it.

**Comment by pfeilbr** [comment]
Ran into the same issue.  An a mac with McAfee installed and a process named `cma` was using `8081`

Used the following to find it.

```
$ sudo lsof -n -i4TCP:8081 | grep LISTEN
```

then

```
$ kill -9 <cma process id>
```

**Comment by vjeux** [comment]
@amasad maybe we should use a different port than 8081 as the default, it seems like it's being used by a lot of things. Maybe 9381

**Comment by amasad** [comment]
@vjeux good point, but it'd be great to have an answer for this regardless (I'm sure someone will run into a problem regardless of the number we choose.

**Comment by JustinCarmony** [comment]
@amasad @vjeux I'd agree. Given how the port is hard coded in several spots, it would mitigate collisions. We can still keep the checking in #308, just change the port.

**Comment by amasad** [comment]
I believe this has been fixed via b6503ba4317f26665e8ae287d1addbe03cd50373
